### PR TITLE
happy birthday hunter

### PR DESCRIPTION
**Pull request by mokshat-drexel**
`user authentication`

### CODE REVIEW

File: src/app/core/services/user_service.py [hunk 1]
```diff
@@ -25,7 +27,7 @@ def get_user(user_id: int, session: ScopedSession):
 
 	return user
 
-def create_user(email: str, session: ScopedSession):
+def create_user(email, username, password, session: ScopedSession):
```
Comment (Exactol) [inline src/app/core/services/user_service.py]: Add type hints

File: src/app/core/services/user_service.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
 from datetime import datetime, tzinfo
+import imp
```
Comment (Exactol) [inline src/app/core/services/user_service.py]: Why is this imported

File: src/app/core/services/user_service.py [hunk 3]
```diff
@@ -52,6 +60,33 @@ def create_user(email: str, session: ScopedSession):
...
+def login(email: str, password: str, session: ScopedSession) -> str | None :
+	"""
+	Args:
+			email (str): registered user's email ID
+			password (str): registered user's password
+	"""
+	#query = select(User).where(User.email == email).execute()
+	
+	# user = session.query(exists(User).where(User.email == email)
+
```
Comment (Exactol) [inline src/app/core/services/user_service.py]: Remove commented code if not used

File: src/app/core/services/user_service.py [hunk 4]
```diff
@@ -52,6 +60,33 @@ def create_user(email: str, session: ScopedSession):
...
+	try:
+		#user: User | None = session.execute(user).scalars().one_or_none()
+		user: User | None = session.query(User).filter(User.email == email).one_or_none()
+		if user is not None:
+			if authentication.check_password(password.encode('utf-8'), user.password_hash):
+				# Generate JWT
+				return authentication.create_jwt(user.username, user.user_id)
+				#return None
+				#return user.user_id
```
Comment (Exactol) [inline src/app/core/services/user_service.py]: Remove commented code

File: src/app/protocols/http/utils/authentication.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    """
+    Authentication function 
+    """
+    return check_jwt_valid(token)
+
+# def decrypt_jwt(token: bytes) -> dict:
+#     """
+#     Decrypts a JWT into a python dictionary object
+#     """
+#     return decode(token, KEY)
```
Comment (Exactol) [inline src/app/protocols/http/utils/authentication.py]: Remove commented code

File: src/app/protocols/http/utils/authentication.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    """
+    Transforms a plain-text password string into a hashed one
+    """
+    return hashpw(unhashed_password, gensalt()).decode('utf-8')
+
+def check_password(password: str, hashed_password: str) -> bool:
+    """
+    Checks if an unhashed password matches its hashed counterpart
+    """
+    return checkpw(password, hashed_password)
```
Comment (Exactol) [inline src/app/protocols/http/utils/authentication.py]: This should probably be in the core

File: src/app/protocols/http/v1/views/user_views.py [hunk 7]
```diff
@@ -39,12 +41,30 @@ def get(self, user_id: int):
...
+	def post(self, body: dict):
+		try:
+			jwt = login(body['email'], body['password'], db.session)
+		except ForbiddenError as e:
+			raise Forbidden(str(e))
+		except NotFoundError as e:
+			raise NotFound(str(e))
+		except Exception as e:
+			raise InternalServerError
+		return jwt, 200
```
Comment (Exactol) [inline src/app/protocols/http/v1/views/user_views.py]: don't need to explicitly return 200, its the default HTTP code

**Review by Exactol — CHANGES_REQUESTED** [review]
Are the other endpoints (plants, devices, etc) also protected?

File: src/app/protocols/http/utils/authentication.py (hunk 6 above)
Comment (mokshat-drexel) [reply src/app/protocols/http/utils/authentication.py]: moved to `authorization.py`

File: src/app/core/services/user_service.py [hunk 8]
```diff
@@ -52,6 +60,28 @@ def create_user(email: str, session: ScopedSession):
...
+
+def login(email: str, password: str, session: ScopedSession) -> str | None :
+	"""
+	Args:
+			email (str): registered user's email ID
+			password (str): registered user's password
+	"""
+
+	try:
+		#user: User | None = session.execute(user).scalars().one_or_none()
```
Comment (Exactol) [inline src/app/core/services/user_service.py]: Remove commented out code

**Review by Exactol — APPROVED** [review]

**Pull request merged**